package io.legado.app.ui.book.read.page

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Typeface
import android.graphics.drawable.LayerDrawable
import android.view.LayoutInflater
import android.widget.FrameLayout
import androidx.core.content.ContextCompat
import androidx.core.graphics.drawable.toDrawable
import androidx.core.net.toUri
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.isGone
import io.legado.app.R
import io.legado.app.constant.AppConst.timeFormat
import io.legado.app.data.entities.Bookmark
import io.legado.app.databinding.ViewBookPageBinding
import io.legado.app.help.config.ReadBookConfig
import io.legado.app.model.ReadBook
import io.legado.app.ui.book.read.page.entities.TextLine
import io.legado.app.ui.book.read.page.entities.TextPage
import io.legado.app.ui.book.read.page.entities.TextPos
import io.legado.app.ui.book.read.page.provider.ChapterProvider
import io.legado.app.ui.config.readConfig.ReadConfig
import io.legado.app.ui.widget.BatteryView
import io.legado.app.utils.activity
import io.legado.app.utils.applyNavigationBarPadding
import io.legado.app.utils.applyStatusBarPadding
import io.legado.app.utils.dpToPx
import io.legado.app.utils.gone
import io.legado.app.utils.isContentScheme
import io.legado.app.utils.setOnApplyWindowInsetsListenerCompat
import io.legado.app.utils.statusBarHeight
import splitties.views.backgroundColor
import java.io.File
import java.util.Date

/**
 * 页面视图
 */
class PageView(
    context: Context,
    callBack: ContentTextView.CallBack? = null,
) : FrameLayout(context) {

    private val binding = ViewBookPageBinding.inflate(LayoutInflater.from(context), this, true)
    private var battery = 100
    private var tvTitle: BatteryView? = null
    private var tvTime: BatteryView? = null
    private var tvBattery: BatteryView? = null
    private var tvBatteryP: BatteryView? = null
    private var tvPage: BatteryView? = null
    private var tvTotalProgress: BatteryView? = null
    private var tvTotalProgress1: BatteryView? = null
    private var tvPageAndTotal: BatteryView? = null
    private var tvBookName: BatteryView? = null
    private var tvTimeBattery: BatteryView? = null
    private var tvTimeBatteryP: BatteryView? = null
    private var tvTitleArrow: BatteryView? = null
    private var tvBatteryInside: BatteryView? = null
    private var tvBatteryIcon: BatteryView? = null
    private var tvBatteryClassic: BatteryView? = null
    private var tvTimeBatteryClassic: BatteryView? = null
    private var tvTitleArrowClassic: BatteryView? = null

    private var isMainView = false
    var isScroll = false

    val headerHeight: Int
        get() {
            val h1 = if (binding.vwStatusBar.isGone) 0 else binding.vwStatusBar.height
            val h2 = if (binding.llHeader.isGone) 0 else binding.llHeader.height
            return h1 + h2 + binding.vwRoot.paddingTop
        }

    val imgBgPaddingStart: Int
        get() {
            return binding.vwRoot.paddingStart
        }

    init {
        callBack?.let { binding.contentTextView.setCallBack(it) }
        upStyle()
        binding.vwStatusBar.applyStatusBarPadding()
        binding.vwNavigationBar.applyNavigationBarPadding()
    }

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)
        upBg()
    }

    fun upStyle() = binding.run {
        upTipStyle()
        ReadBookConfig.let {
            val textColor = it.textColor
            val headerColor = with(ReadBookConfig) {
                if (tipHeaderColor == 0) textColor else tipHeaderColor
            }
            val footerColor = with(ReadBookConfig) {
                if (tipFooterColor == 0) textColor else tipFooterColor
            }
            val tipDividerColor = with(ReadBookConfig) {
                when (tipDividerColor) {
                    -1 -> ContextCompat.getColor(context, R.color.divider)
                    0 -> textColor
                    else -> tipDividerColor
                }
            }
            tvHeaderLeft.setColor(headerColor)
            tvHeaderMiddle.setColor(headerColor)
            tvHeaderRight.setColor(headerColor)
            tvFooterLeft.setColor(footerColor)
            tvFooterMiddle.setColor(footerColor)
            tvFooterRight.setColor(footerColor)
            vwTopDivider.backgroundColor = tipDividerColor
            vwBottomDivider.backgroundColor = tipDividerColor
            upStatusBar()
            upNavigationBar()
            upPaddingDisplayCutouts()
            llHeader.setPadding(
                it.headerPaddingLeft.dpToPx(),
                it.headerPaddingTop.dpToPx(),
                it.headerPaddingRight.dpToPx(),
                it.headerPaddingBottom.dpToPx()
            )
            llFooter.setPadding(
                it.footerPaddingLeft.dpToPx(),
                it.footerPaddingTop.dpToPx(),
                it.footerPaddingRight.dpToPx(),
                it.footerPaddingBottom.dpToPx()
            )
            vwTopDivider.gone(llHeader.isGone || !it.showHeaderLine)
            vwBottomDivider.gone(llFooter.isGone || !it.showFooterLine)
        }
        upTime()
        upBattery(battery)
    }

    private fun loadTypeface(fontPath: String): Typeface? {
        return runCatching {
            when {
                fontPath.isContentScheme() -> {
                    context.contentResolver
                        .openFileDescriptor(fontPath.toUri(), "r")
                        ?.use {
                            Typeface.Builder(it.fileDescriptor).build()
                        }
                }
                fontPath.isNotEmpty() -> {
                    Typeface.Builder(File(fontPath)).build()
                }
                else -> null
            }
        }.getOrNull()
    }

    /**
     * 显示状态栏时隐藏header
     */
    fun upStatusBar() = with(binding.vwStatusBar) {
        setPadding(paddingLeft, context.statusBarHeight, paddingRight, paddingBottom)
        isGone = ReadBookConfig.hideStatusBar || activity?.isInMultiWindowMode == true
    }

    fun upNavigationBar() {
        binding.vwNavigationBar.isGone = ReadBookConfig.hideNavigationBar
    }

    fun upPaddingDisplayCutouts() {
        if (ReadConfig.paddingDisplayCutouts) {
            binding.vwRoot.setOnApplyWindowInsetsListenerCompat { _, windowInsets ->
                val insets = windowInsets.getInsets(WindowInsetsCompat.Type.displayCutout())
                binding.vwRoot.setPadding(
                    insets.left,
                    if (binding.vwStatusBar.isGone) insets.top else 0,
                    insets.right,
                    insets.bottom
                )
                windowInsets
            }
        } else {
            ViewCompat.setOnApplyWindowInsetsListener(binding.vwRoot, null)
            binding.vwRoot.setPadding(0, 0, 0, 0)
        }
    }

    /**
     * 更新阅读信息
     */
    private fun upTipStyle() = binding.run {
        tvHeaderLeft.tag = null
        tvHeaderMiddle.tag = null
        tvHeaderRight.tag = null
        tvFooterLeft.tag = null
        tvFooterMiddle.tag = null
        tvFooterRight.tag = null
        llHeader.isGone = when (ReadBookConfig.headerMode) {
            1 -> false
            2 -> true
            else -> !ReadBookConfig.hideStatusBar
        }
        llFooter.isGone = when (ReadBookConfig.footerMode) {
            1 -> true
            else -> false
        }
        ReadBookConfig.apply {
            tvHeaderLeft.isGone = tipHeaderLeft == tipNone
            tvHeaderMiddle.isGone = tipHeaderMiddle == tipNone
            if (tipHeaderRight == tipNone) {
                if (tipHeaderMiddle == tipNone && tipHeaderLeft == tipNone) {
                    tvHeaderRight.isGone = true
                } else {
                    tvHeaderRight.isGone = false
                    tvHeaderRight.batteryMode = BatteryView.BatteryMode.EMPTY
                }
            } else {
                tvHeaderRight.isGone = false
                tvHeaderRight.batteryMode = BatteryView.BatteryMode.NO_BATTERY
            }
            tvFooterLeft.isGone = tipFooterLeft == tipNone
            tvFooterMiddle.isGone = tipFooterMiddle == tipNone
            if (tipFooterRight == tipNone) {
                if (tipFooterLeft == tipNone && tipFooterMiddle == tipNone) {
                    tvFooterRight.isGone = true
                } else {
                    tvFooterRight.isGone = false
                    tvFooterRight.batteryMode = BatteryView.BatteryMode.EMPTY
                }
            } else {
                tvFooterRight.isGone = false
                tvFooterRight.batteryMode = BatteryView.BatteryMode.NO_BATTERY
            }
        }
        val tipTypeface = loadTypeface(ReadBookConfig.headerFont) ?: ChapterProvider.typeface
        val tipTextSize = ReadBookConfig.headerFontSize.toFloat()
        tvTitle = getTipView(ReadBookConfig.tipChapterTitle)?.apply {
            tag = ReadBookConfig.tipChapterTitle
            typeface = tipTypeface
            textSize = tipTextSize
            batteryMode = BatteryView.BatteryMode.NO_BATTERY
        }
        tvTitleArrow = getTipView(ReadBookConfig.tipChapterTitleArrow)?.apply {
            tag = ReadBookConfig.tipChapterTitleArrow
            typeface = Typeface.DEFAULT
            textSize = tipTextSize
            batteryMode = BatteryView.BatteryMode.ARROW
        }
        tvTitleArrowClassic = getTipView(ReadBookConfig.tipChapterTitleArrowClassic)?.apply {
            tag = ReadBookConfig.tipChapterTitleArrowClassic
            typeface = tipTypeface
            textSize = tipTextSize
            batteryMode = BatteryView.BatteryMode.ARROW
        }
        tvTime = getTipView(ReadBookConfig.tipTime)?.apply {
            tag = ReadBookConfig.tipTime
            typeface = tipTypeface
            textSize = tipTextSize
            batteryMode = BatteryView.BatteryMode.NO_BATTERY
        }
        tvBattery = getTipView(ReadBookConfig.tipBattery)?.apply {
            tag = ReadBookConfig.tipBattery
            typeface = Typeface.DEFAULT
            textSize = tipTextSize
            batteryMode = BatteryView.BatteryMode.OUTER
        }
        tvBatteryClassic = getTipView(ReadBookConfig.tipBatteryClassic)?.apply {
            tag = ReadBookConfig.tipBatteryClassic
            textSize = tipTextSize
            batteryMode = BatteryView.BatteryMode.CLASSIC
        }
        tvBatteryInside = getTipView(ReadBookConfig.tipBatteryInside)?.apply {
            tag = ReadBookConfig.tipBatteryInside
            typeface = Typeface.DEFAULT
            textSize = tipTextSize
            batteryMode = BatteryView.BatteryMode.INNER
        }
        tvBatteryIcon = getTipView(ReadBookConfig.tipBatteryIcon)?.apply {
            tag = ReadBookConfig.tipBatteryIcon
            typeface = Typeface.DEFAULT
            textSize = tipTextSize
            batteryMode = BatteryView.BatteryMode.ICON
        }
        tvPage = getTipView(ReadBookConfig.tipPage)?.apply {
            tag = ReadBookConfig.tipPage
            typeface = tipTypeface
            textSize = tipTextSize
            batteryMode = BatteryView.BatteryMode.NO_BATTERY
        }
        tvTotalProgress = getTipView(ReadBookConfig.tipTotalProgress)?.apply {
            tag = ReadBookConfig.tipTotalProgress
            batteryMode = BatteryView.BatteryMode.NO_BATTERY
            typeface = tipTypeface
            textSize = tipTextSize
        }
        tvTotalProgress1 = getTipView(ReadBookConfig.tipTotalProgress1)?.apply {
            tag = ReadBookConfig.tipTotalProgress1
            batteryMode = BatteryView.BatteryMode.NO_BATTERY
            typeface = tipTypeface
            textSize = tipTextSize
        }
        tvPageAndTotal = getTipView(ReadBookConfig.tipPageAndTotal)?.apply {
            tag = ReadBookConfig.tipPageAndTotal
            batteryMode = BatteryView.BatteryMode.NO_BATTERY
            typeface = tipTypeface
            textSize = tipTextSize
        }
        tvBookName = getTipView(ReadBookConfig.tipBookName)?.apply {
            tag = ReadBookConfig.tipBookName
            batteryMode = BatteryView.BatteryMode.NO_BATTERY
            typeface = tipTypeface
            textSize = tipTextSize
        }
        tvTimeBattery = getTipView(ReadBookConfig.tipTimeBattery)?.apply {
            tag = ReadBookConfig.tipTimeBattery
            typeface = Typeface.DEFAULT
            textSize = tipTextSize
            batteryMode = BatteryView.BatteryMode.TIME
        }
        tvTimeBatteryClassic = getTipView(ReadBookConfig.tipTimeBatteryClassic)?.apply {
            tag = ReadBookConfig.tipTimeBatteryClassic
            typeface = tipTypeface
            textSize = tipTextSize
            batteryMode = BatteryView.BatteryMode.CLASSIC
        }
        tvBatteryP = getTipView(ReadBookConfig.tipBatteryPercentage)?.apply {
            tag = ReadBookConfig.tipBatteryPercentage
            batteryMode = BatteryView.BatteryMode.NO_BATTERY
            typeface = tipTypeface
            textSize = tipTextSize
        }
        tvTimeBatteryP = getTipView(ReadBookConfig.tipTimeBatteryPercentage)?.apply {
            tag = ReadBookConfig.tipTimeBatteryPercentage
            batteryMode = BatteryView.BatteryMode.NO_BATTERY
            typeface = tipTypeface
            textSize = tipTextSize
        }
    }

    /**
     * 获取信息视图
     * @param tip 信息类型
     */
    private fun getTipView(tip: Int): BatteryView? = binding.run {
        return when (tip) {
            ReadBookConfig.tipHeaderLeft -> tvHeaderLeft
            ReadBookConfig.tipHeaderMiddle -> tvHeaderMiddle
            ReadBookConfig.tipHeaderRight -> tvHeaderRight
            ReadBookConfig.tipFooterLeft -> tvFooterLeft
            ReadBookConfig.tipFooterMiddle -> tvFooterMiddle
            ReadBookConfig.tipFooterRight -> tvFooterRight
            else -> null
        }
    }

    /**
     * 更新背景
     */
    fun upBg() {
        binding.vwRoot.background = LayerDrawable(
            arrayOf(
                ReadBookConfig.bgMeanColor.toDrawable(),
                ReadBookConfig.bg
            )
        )
        upBgAlpha()
    }

    /**
     * 更新背景透明度
     */
    fun upBgAlpha() {
        ReadBookConfig.bg?.alpha = (ReadBookConfig.bgAlpha / 100f * 255).toInt()
        binding.vwRoot.invalidate()
    }

    /**
     * 更新时间信息
     */
    fun upTime() {
        tvTime?.text = timeFormat.format(Date(System.currentTimeMillis()))
        upTimeBattery()
    }

    /**
     * 更新电池信息
     */
    @SuppressLint("SetTextI18n")
    fun upBattery(battery: Int) {
        this.battery = battery
        tvBattery?.setBattery(battery)
        tvBatteryClassic?.setBattery(battery)
        tvBatteryInside?.setBattery(battery)
        tvBatteryIcon?.setBattery(battery)
        tvBatteryP?.text = "$battery%"
        upTimeBattery()
    }

    /**
     * 更新电池信息
     */
    @SuppressLint("SetTextI18n")
    private fun upTimeBattery() {
        val time = timeFormat.format(Date(System.currentTimeMillis()))
        tvTimeBattery?.setBattery(battery, time)
        tvTimeBatteryP?.setBattery(battery, time)
        tvTimeBatteryClassic?.setBattery(battery, time)
        tvTimeBatteryClassic?.text = "$time $battery%"
        tvTimeBatteryP?.text = "$time $battery%"
    }

    /**
     * 设置内容
     */
    fun setContent(textPage: TextPage, resetPageOffset: Boolean = true) {
        if (isMainView && !isScroll) {
            setProgress(textPage)
        } else {
            post {
                setProgress(textPage)
            }
        }
        if (resetPageOffset) {
            resetPageOffset()
        }
        binding.contentTextView.setContent(textPage)
    }

    fun invalidateContentView() {
        binding.contentTextView.invalidate()
    }

    /**
     * 设置无障碍文本
     */
    fun setContentDescription(content: String) {
        binding.contentTextView.contentDescription = content
    }

    /**
     * 重置滚动位置
     */
    fun resetPageOffset() {
        binding.contentTextView.resetPageOffset()
    }

    /**
     * 设置进度
     */
    @SuppressLint("SetTextI18n")
    fun setProgress(textPage: TextPage) = textPage.apply {
        tvBookName?.setTextIfNotEqual(ReadBook.book?.name)
        tvTitle?.setTextIfNotEqual(textPage.title)
        tvTitleArrow?.setTextIfNotEqual(textPage.title)
        tvTitleArrowClassic?.setTextIfNotEqual(textPage.title)
        val readProgress = readProgress
        tvTotalProgress?.setTextIfNotEqual(readProgress)
        tvTotalProgress1?.setTextIfNotEqual("${chapterIndex.plus(1)}/${chapterSize}")
        if (textChapter.isCompleted) {
            tvPageAndTotal?.setTextIfNotEqual("${index.plus(1)}/$pageSize  $readProgress")
            tvPage?.setTextIfNotEqual("${index.plus(1)}/$pageSize")
        } else {
            val pageSizeInt = pageSize
            val pageSize = if (pageSizeInt <= 0) "-" else "~$pageSizeInt"
            tvPageAndTotal?.setTextIfNotEqual("${index.plus(1)}/$pageSize  $readProgress")
            tvPage?.setTextIfNotEqual("${index.plus(1)}/$pageSize")
        }
    }

    fun setAutoPager(autoPager: AutoPager?) {
        binding.contentTextView.setAutoPager(autoPager)
    }

    fun submitRenderTask() {
        binding.contentTextView.submitRenderTask()
    }

    fun setIsScroll(value: Boolean) {
        isScroll = value
        binding.contentTextView.setIsScroll(value)
    }

    /**
     * 滚动事件
     */
    fun scroll(offset: Int) {
        binding.contentTextView.scroll(offset)
    }

    /**
     * 更新是否开启选择功能
     */
    fun upSelectAble(selectAble: Boolean) {
        binding.contentTextView.selectAble = selectAble
    }

    /**
     * 优先处理页面内单击
     * @return true:已处理, false:未处理
     */
    fun onClick(x: Float, y: Float): Boolean {
        return binding.contentTextView.click(x, y - headerHeight)
    }

    /**
     * 长按事件
     */
    fun longPress(
        x: Float, y: Float,
        select: (textPos: TextPos) -> Unit,
    ) {
        return binding.contentTextView.longPress(x, y - headerHeight, select)
    }

    /**
     * 选择文本
     */
    fun selectText(
        x: Float, y: Float,
        select: (textPos: TextPos) -> Unit,
    ) {
        return binding.contentTextView.selectText(x, y - headerHeight, select)
    }

    fun getCurVisiblePage(): TextPage {
        return binding.contentTextView.getCurVisiblePage()
    }

    fun getReadAloudPos(): Pair<Int, TextLine>? {
        return binding.contentTextView.getReadAloudPos()
    }

    fun markAsMainView() {
        isMainView = true
        binding.contentTextView.isMainView = true
    }

    fun selectStartMove(x: Float, y: Float) {
        binding.contentTextView.selectStartMove(x, y - headerHeight)
    }

    fun selectStartMoveIndex(
        relativePagePos: Int,
        lineIndex: Int,
        charIndex: Int
    ) {
        binding.contentTextView.selectStartMoveIndex(relativePagePos, lineIndex, charIndex)
    }

    fun selectStartMoveIndex(textPos: TextPos) {
        binding.contentTextView.selectStartMoveIndex(textPos)
    }

    fun selectEndMove(x: Float, y: Float) {
        binding.contentTextView.selectEndMove(x, y - headerHeight)
    }

    fun selectEndMoveIndex(
        relativePagePos: Int,
        lineIndex: Int,
        charIndex: Int
    ) {
        binding.contentTextView.selectEndMoveIndex(relativePagePos, lineIndex, charIndex)
    }

    fun selectEndMoveIndex(textPos: TextPos) {
        binding.contentTextView.selectEndMoveIndex(textPos)
    }

    fun getReverseStartCursor(): Boolean {
        return binding.contentTextView.reverseStartCursor
    }

    fun getReverseEndCursor(): Boolean {
        return binding.contentTextView.reverseEndCursor
    }

    fun isLongScreenShot(): Boolean {
        return binding.contentTextView.longScreenshot
    }

    fun resetReverseCursor() {
        binding.contentTextView.resetReverseCursor()
    }

    fun cancelSelect(clearSearchResult: Boolean = false) {
        binding.contentTextView.cancelSelect(clearSearchResult)
    }

    fun createBookmark(): Bookmark? {
        return binding.contentTextView.createBookmark()
    }

    fun relativePage(relativePagePos: Int): TextPage {
        return binding.contentTextView.relativePage(relativePagePos)
    }

    val textPage get() = binding.contentTextView.textPage

    val selectedText: String get() = binding.contentTextView.getSelectedText()

    val selectStartPos get() = binding.contentTextView.selectStart
}
